_posts/2023-03-26-Kafka-IaC.md

```diff
@@ -0,0 +1,249 @@
+---
+layout: post
+title:  "Kafka Infrastructure as Code(IaC)"
+author: dasasathyan
+categories: [ Platform Engineering, Data, Infrastructure, Kafka ]
+featured: true
+hidden: true
+teaser: Provision Kafka Infrastructure with Juli-Ops, Pulumi, Terraform
+toc: true
+---
+
+# Infrastructure as Code(IaC)
+
+Organizations need to automate infrastructure provisioning. On the other hand, those tools can’t be too restrictive on development teams. Development teams should have as much autonomy as possible because that's usually how developers get their best work done. It is not easy to manage Infrastructure as going wrong in any of the steps worsens the situation. Also, replicating the same set of configurations for an additional cluster is quite tedious. Here comes the Infrastructure as Code(IaC) handy. It helps us to provision the same Infrastructure across environments with the help of a source code. The advantages of using an IaC are
+1. Very less time to provision infrastructure
+1. Low risk of Human errors
+1. Idempotency
+1. Reduced configuration steps
+1. Eliminate configuration drift
+
+A few of the IaC tools are Terraform, Pulumi etc.
+
+Apache Kafka is a real-time data streaming technology capable of handling trillions of events. It is a distributed system with servers and clients that communicate via a TCP network protocol. A couple of terminologies to keep in mind are:
+
+1. Brokers - Brokers are servers in Kafka that store event streams from various sources. A Kafka cluster is typically comprised of several brokers. Every broker in a cluster is also a bootstrap server, meaning if you can connect to one broker in a cluster, you can connect to every broker.
+
+1. Topics - The data is written by many processes called produces and the same are read by consumers. The data are partitioned into different partitions called topics. Kafka runs on a cluster of one or more servers called brokers and the partitions are distributed across the cluster. 
+
+1. Kafka Connect - Messages can be copied to and from external applications and data systems with Kafka Connect. There are 2 different types of connectors. They are Source connector and Sink Connectors.
+
+All the above-mentioned infrastructure like Topics, Connectors etc can be configured with IaC tools like julie-ops, terraform, pulumi.
```

Comment:
Need to talk more about why we are comparing JulieOps, Terraform and Pulumi. Also some words about the basis for our comparison